mygrad/computational_graph.py

```diff
@@ -0,0 +1,107 @@
+    """ Builds and renders a computational graph.
+
+        Parameters
+        ----------
+        fin : Tensor
+            The tensor object that will be the final node in the
+            computational graph.
+
+        names : Optional[Dict[str, Tensor]]
+            A dictionary that maps names of Tensors to Tensor objects. If
+            an argument is passed to names, the key name that maps to a Tensor
+            included in the computational graph will be used as a label for the
+            Tensor's node. If no argument is passed, the nodes on the
+            computational graph will display the full Tensor.
+
+            To use the names assigned in the local environment,
+            pass `names=locals()` to the build_graph function.
+
+            If different names are used from the local environment,
+            the key must map to the exact Tensor object. A new Tensor or copy
+            of the original Tensor should not be created as the value in the
+            dictionary.
+
+        render : bool, optional (default=True)
+            If True, build_graph will return a graphviz Digraph object that,
+            when called, will render the computational graph in a Jupyter
+            notebook or the Jupyter Qt console. If False, nothing is returned.
+
+        save : bool, optional (default=False)
+            If True, build_graph will save a rendered computational graph to
+            the current working directory as `computational_graph.pdf`.
+
+        dims : bool, optional (default=False)
+            If True, Tensor dimensions are added to Node labels.
+
+        dtypes : bool, optional (default=False)
+            If True, Tensor data types are added to Node labels.
+
+        sum_stats : bool, optional (default=False)
+            If True, Tensor minimums, maximums, medians, and means are
+            added to Node labels.
+
+        Returns
+        -------
+        graphviz.Digraph
+
+        Notes
+        -----
+        build_graph requires that Graphviz is installed.
+    """
+    assert isinstance(fin, Tensor), "fin must be a Tensor"
+    assert isinstance(names, (dict, type(None)))
+    assert isinstance(render, bool)
+    assert isinstance(save, bool)
+
+    graph = Digraph()
+    graph.node_attr.update(fontsize="12")
+
+    _add_node(fin, graph, **{'names':names, 'dims':dims, 'dtypes':dtypes, 'sum_stats':sum_stats})
+
+    if save:
+        graph.render(filename="computational_graph", cleanup=True)
+
+    if render:
+        return graph
+
+
+def _add_node(node, graph, op_id=None, **kwargs):
+    """ Recursively traces computational graph and adds nodes to Digraph. """
+    node_lab = node.__repr__()
+    if kwargs['names'] is not None:
+        for key in kwargs['names']:
+            if id(kwargs['names'][key]) == id(node):
+                node_lab = key
+    if kwargs['dims']:
+        node_lab = node_lab + "\nDims: " + str(node.shape)
```

Comment:
```python
"\nDims: " + str(node.shape)
```
is better written as:
```python
"\nDims: {}".format(node.shape)
```
This is especially the case for longer strings. You can do named formatted fields:

```python
"{apple}, {moo}, {cat}".format(apple=2, moo="hi", cat=True)
```